File: solutions/034.Search_for_a_Range/AC_binary_search_logn.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_binary_search_logn.java
*  Create Date: 2015-03-13 16:25:37
*  Descripton:   
*/

import java.util.*;

public class Solution {

    public int[] searchRange(int[] A, int target) {
        int[] ret = new int[2];
        ret[0] = ret[1] = -1;
        int left = 0, right = A.length - 1, mid;

        while (left <= right) {
            if (A[left] == target && A[right] == target) {
                ret[0] = left;
                ret[1] = right;
                break;
            }

            mid = (right + left) / 2;
            if (A[mid] < target) {
                left = mid + 1;
            } else if (A[mid] > target) {
                right = mid - 1;
            } else {
                if (A[right] == target) {
                    ++left;
                } else {
                    --right;
                }
            }
        }

        return ret;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 3};
        System.out.println(s.searchRange(input, 1));
    }
}

File: solutions/153.Find_Minimum_in_Rotated_Sorted_Array/AC_binary_search_logn.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_binary_search_logn.cpp
*  Create Date: 2015-03-18 15:16:39
*  Descripton:  
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    int findMin(vector<int> &num) {
        int left = 0, right = num.size() - 1;
        int mid;
        while (left <= right) {
            mid = (left + right) / 2;
            if (num[left] <= num[mid] && num[mid] <= num[right]) {
                return num[left];
            } else if (num[left] <= num[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return num[0];
    }
};

int main() {
    vector<int> num;
    num.push_back(3);
    num.push_back(1);
    num.push_back(2);
    Solution s;
    cout << s.findMin(num) << endl;
    return 0;
}

